src/Hazelcast.Net/Sql/ISqlQueryResult.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Hazelcast.Sql
+{
+    /// <summary>
+    /// Represents SQL query (SELECT ...).
+    /// Provides methods to enumerate queried rows or cancel/dispose the query.
+    /// </summary>
+    public interface ISqlQueryResult: IAsyncEnumerator<SqlRow>, IAsyncDisposable
+    {
+        /// <summary>
+        /// <para>
+        /// Creates a one-off <see cref="IAsyncEnumerable{T}"/> around <see cref="SqlRow"/>s in row set.
+        /// </para>
+        /// <para>
+        /// Reusing obtained enumerable or invoking this method again
+        /// will continue previous enumeration instead of starting a new one.
+        /// </para>
+        /// </summary>
+        /// <param name="cancellationToken">Optional <see cref="CancellationToken"/> to cancel enumeration in progress.</param>
+        /// <exception cref="ObjectDisposedException">SQL result was disposed.</exception>
+        IAsyncEnumerable<SqlRow> EnumerateOnceAsync(CancellationToken cancellationToken = default);
```

Comment:
We don't want the `CancellationToken` here. The .NET way with `IAsyncEnumerable<>` is to do
```
await foreach (var row in result.EnumerateOnceAsync().WithCancellation(cancellationToken)) {...}
```
And the `cancellationToken` is passed to the enumerator.